client/client_test.go

```diff
@@ -1354,6 +1354,352 @@ func TestSubscribePredefined(t *testing.T) {
 	}
 }
 
+func TestUnsubscribe(t *testing.T) {
+	assert := assert.New(t)
+
+	clientID := "test-client"
+	topic := "test/a"
+	qos := uint8(0)
+
+	stp := newTestSetup(t, clientID)
+	defer stp.cancel()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		stp.connect(clientID)
+
+		// client --SUBSCRIBE--> GW
+		subscribe := stp.recv().(*msgs.SubscribeMessage)
+		subscribe.QOS = qos
+		assert.Equal([]byte(topic), subscribe.TopicName)
+		assert.Equal(msgs.TIT_STRING, subscribe.TopicIDType)
+
+		// client <--SUBACK-- GW
+		suback := msgs.NewSubackMessage(1, 0, msgs.RC_ACCEPTED)
+		suback.CopyMessageID(subscribe)
+		stp.send(suback)
+
+		// client <--PUBLISH-- GW
+		msgID := uint16(123)
+		publish := msgs.NewPublishMessage(suback.TopicID,
+			msgs.TIT_REGISTERED, []byte(""), qos, false, false)
+		publish.SetMessageID(msgID)
+		stp.send(publish)
+
+		// client --UNSUBSCRIBE--> GW
+		unsubscribe := stp.recv().(*msgs.UnsubscribeMessage)
+		assert.Equal([]byte(topic), unsubscribe.TopicName)
+		assert.Equal(msgs.TIT_STRING, unsubscribe.TopicIDType)
+
+		// client <--UNSUBACK-- GW
+		unsuback := msgs.NewUnsubackMessage()
+		unsuback.CopyMessageID(unsubscribe)
+		stp.send(unsuback)
+
+		stp.disconnect()
+	}()
+
+	if err := stp.client.Connect(); err != nil {
+		stp.t.Fatal(err)
+	}
+	assert.Equal(util.StateActive, stp.client.state.Get())
+
+	callbackFired := make(chan struct{})
+
+	callback := func(client *Client, topic string, msg *msgs.PublishMessage) {
+		close(callbackFired)
+	}
+	if err := stp.client.Subscribe(topic, qos, callback); err != nil {
+		stp.t.Fatal(err)
+	}
+
+	select {
+	case <-callbackFired:
+		// ok
+	case <-time.After(time.Second):
+		stp.t.Fatal("subscribe callback not fired")
+	}
+
+	if err := stp.client.Unsubscribe(topic); err != nil {
+		stp.t.Fatal(err)
+	}
+
+	if err := stp.client.Disconnect(); err != nil {
+		stp.t.Fatal(err)
+	}
+	assert.Equal(util.StateDisconnected, stp.client.state.Get())
+	stp.assertClientDone()
+
+	wg.Wait()
+}
+
+func TestUnsubscribeWildcard(t *testing.T) {
```

Comment:
This test doesn't make sense to me. The most of it tests `PUBLISH` handling.

We can discuss it in detail elsewhere.